### DISCUSSION TITLE
Help redirect http to https

### DISCUSSION THREAD

**Discussion by septiyadi396** [adonisjs/core#4021] [Help] [answered]
Hello, my website is using HTTPS and was successfully deployed from a build. However, in some cases, some of my users are unable to access it if they type the URL without HTTPS. Instead of redirecting to https://myawesomeapp.site, it becomes http://myawesome.site. My team, which includes one lecturer, thinks the problem is that they are using an old version of the browser. While we don't think asking users to update their browser is a solution, we are wondering if there is a solution from the code or server side.
Some devices that we have found to not work include an old MacBook Pro (Safari), an old iPhone (Safari), and an Acer/Asus laptop (Windows, forgot the browser).

**server.ts**
```
import 'reflect-metadata'
import sourceMapSupport from 'source-map-support'
import { Ignitor } from '@adonisjs/core/build/standalone'
import { createServer } from "https";
import { join } from 'path';
import { readFileSync } from 'fs';

sourceMapSupport.install({ handleUncaughtExceptions: false })

new Ignitor(__dirname)
  .httpServer()
  .start((handle) => {
    return createServer(
      {
        key: readFileSync(join(__dirname + '/ssl_certificates/unsyiah-usk.key'), 'utf8'),
        cert: readFileSync(join(__dirname + '/ssl_certificates/unsyiah-usk.crt'), 'utf8'),
      },
      handle
    )
  })
```

I also tried using middleware, and it works if HTTPS is used. However, if HTTP is used, it is not redirecting.

**app/Middleware/RedirectToHttp.ts**
```
import type { HttpContextContract } from '@ioc:Adonis/Core/HttpContext'

export default class RedirectToHttp {
  public async handle({request, response}: HttpContextContract, next: () => Promise<void>) {
    if (request.secure()) {
      return await next()
    }

    return response.redirect(`https://${request.hostname()}${request.originalUrl()}`)
  }
}
```

sorry if there is a wrong word, i use translator

**Comment by Melchyore** [comment, marked as answer]
@septiyadi396 hello, why are you not using Apache or Nginx to redirect URLs?

What you are doing now is redirecting users just before sending the responses, while you need to do it before the requests hit your application.

**Reply by septiyadi396** [reply]
I don't have access to the server, only the lecturer has access. And he use [pm2](https://pm2.io), not apache or nginx.

I just read about nginx reverse proxy for node js, if I use nginx can somehow it redirect http to https? (he's a bit of a perfectionist, so it should be made able to redirect to https)
And if nginx handle certification thing, should the code back to http only and remove that middleware?

This actually my first project, sorry for many basic questions.
Thanks for responding

**Reply by Barbapapazes** [reply]
Hello,

Nginx is used as a reverse proxy. So, it handles request from the outside and pass it inside your server. Request from client to nginx uses https with certificate provided by let's encryte (and the certbot) for exemple. 

But clearly, avoid Nodejs to handle https and to expose it to real world. 

https://medium.com/intrinsic-blog/why-should-i-use-a-reverse-proxy-if-node-js-is-production-ready-5a079408b2ca 


https://en.m.wikipedia.org/wiki/Reverse_proxy
